runtime/compiler/env/VMJ9Server.cpp

```diff
@@ -2434,6 +2434,47 @@ TR_J9ServerVM::isPortableSCCEnabled()
    return J9_ARE_ANY_BITS_SET(vmInfo->_extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_ENABLE_PORTABLE_SHARED_CACHE);
    }
 
+bool
+TR_J9ServerVM::inSnapshotMode()
+   {
+   JITServer::ServerStream *stream = _compInfoPT->getMethodBeingCompiled()->_stream;
+   auto *vmInfo = _compInfoPT->getClientData()->getOrCacheVMInfo(stream);
+   if (vmInfo->_isSnapshotModeEnabled)
+      {
+      if (vmInfo->_isNonPortableRestoreMode)
+         {
+         // The VM starts in snapshot mode, but after a restore it will exit snapshot mode
+         if (vmInfo->_inSnapshotMode)
+            {
+            // Must ask confirmation from client that it is still running in snapshot mode
+            stream->write(JITServer::MessageType::VM_inSnapshotMode, JITServer::Void());
+            vmInfo->_inSnapshotMode = std::get<0>(stream->read<bool>()); // cache the response
+            return vmInfo->_inSnapshotMode;
+            }
+         else // Once false, it will never be true again
+            {
+            return false;
+            }
```

Comment:
When a client disconnects from the server, does the server still hold on to the `vmInfo` for some time or does it remove it immediately? If it's the latter, then we don't need to always ask the client if it's in snapshot mode because all client compilation threads are suspended prior to checkpoint, so all connections should be severed at that point.

Reply:
`vmInfo` is deleted if the client JVM shuts down nicely and has the chance of sending a termination message to the server. Otherwise, the server deletes the vmInfo if it doesn't receive any message from the client for 90 minutes.

Reply:
To be more precise, just because the client has closed all connections to the server, does not make the server drop the vmInfo.